### PR TITLE
added Curve in state machine interpolation

### PR DESCRIPTION
**Pull request by jeronimo-schreyer**
added Curve in state machine interpolation for better control over blending
theres a proposal for this https://github.com/godotengine/godot-proposals/issues/1402

here you can see the interpolation with and without curve setting

https://user-images.githubusercontent.com/29764541/180624853-6a53dd8a-a6ee-4a04-8547-a3f9656fbcf8.mp4

### CODE REVIEW

**Comment by Calinou** [comment]
@jeronimo-schreyer Remember that for GitHub video previews to work, you need to have a blank line before *and* after the video URL. I edited your post accordingly, but remember to do this in the future :slightly_smiling_face:

**Comment by jeronimo-schreyer** [comment]
I didn't know why I could see it OK in the preview and not in the final posted version. Thanks.

This build error says I need to add documentation for the new Curve property, right?

**Comment by Calinou** [comment]
> This build error says I need to add documentation for the new Curve property, right?

Yes, and make sure you're using tab indentation within the XML instead of spaces (except within `[codeblock]`).

**Comment by jeronimo-schreyer** [comment]
I'm almost certain I use tabs instead of spaces. Is there a way to be sure?

**Comment by TokageItLab** [comment]
IMO it is good, it could be implemented also for NodeTransition, but maybe after this is Approved, we can send additional PRs.

**Comment by jeronimo-schreyer** [comment]
I added this for NodeTransition also on my latest commit

File: scene/animation/animation_blend_tree.cpp [hunk 1]
```diff
@@ -734,26 +742,27 @@ double AnimationNodeTransition::process(double p_time, bool p_seek, bool p_seek_
...
-		if (inputs[current].auto_advance && rem <= xfade) {
+		if (inputs[current].auto_advance && rem <= xfade_time) {
 			set_parameter(this->current, (current + 1) % enabled_inputs);
 		}
 
 	} else { // cross-fading from prev to current
 
-		float blend = xfade == 0 ? 0 : (prev_xfading / xfade);
+		float blend = xfade_time == 0 ? 0 : (prev_xfading / xfade_time);
+		float curved_blend = (xfade_curve.is_valid()) ? xfade_curve->interpolate(1.0 - blend) : 1.0 - blend;
```
Comment (TokageItLab) [inline scene/animation/animation_blend_tree.cpp:752]:
```suggestion
		float curved_blend = (xfade_curve.is_valid()) ? 1.0 - xfade_curve->interpolate(blend) : 1.0 - blend;
```

Comment (TokageItLab) [reply scene/animation/animation_blend_tree.cpp:752]: The interpolation calculation is wrong.

File: scene/animation/animation_node_state_machine.cpp [hunk 2]
```diff
@@ -420,10 +432,12 @@ double AnimationNodeStateMachinePlayback::process(AnimationNodeStateMachine *p_s
 		}
 	}
 
-	float rem = p_state_machine->blend_node(current, p_state_machine->states[current].node, p_time, p_seek, p_seek_root, fade_blend, AnimationNode::FILTER_IGNORE, true);
+	float curved_blend = (current_curve.is_valid()) ? current_curve->interpolate(fade_blend) : fade_blend;
+	float rem = p_state_machine->blend_node(current, p_state_machine->states[current].node, p_time, p_seek, p_seek_root, curved_blend, AnimationNode::FILTER_IGNORE, true);
 
 	if (fading_from != StringName()) {
-		p_state_machine->blend_node(fading_from, p_state_machine->states[fading_from].node, p_time, p_seek, p_seek_root, 1.0 - fade_blend, AnimationNode::FILTER_IGNORE, true);
+		curved_blend = (current_curve.is_valid()) ? current_curve->interpolate(1.0 - fade_blend) : 1.0 - fade_blend;
```
Comment (TokageItLab) [inline scene/animation/animation_node_state_machine.cpp:439]:
```suggestion
		curved_blend = (current_curve.is_valid()) ? 1.0 - current_curve->interpolate(fade_blend) : 1.0 - fade_blend;
```
Same above.

**Review by TokageItLab — CHANGES_REQUESTED** [review]
You must follow the rebasing work-flow recommended by Godot. See [Pull request workflow](https://docs.godotengine.org/en/stable/community/contributing/pr_workflow.html#the-interactive-rebase).

In brief, check out `74d84d170aab4e692a81165872f040296a2ca7f0` and apply the fixes, then do a `--fixup`. In the end, there should be only one commit.

File: scene/animation/animation_blend_tree.cpp [hunk 3]
```diff
@@ -734,26 +742,27 @@ double AnimationNodeTransition::process(double p_time, bool p_seek, bool p_seek_
...
 		}
 
+		curved_blend = (xfade_curve.is_valid()) ? xfade_curve->interpolate(blend) : blend;
 		if (p_seek) {
-			blend_input(prev, p_time, true, p_seek_root, blend, FILTER_IGNORE, true);
+			blend_input(prev, p_time, true, p_seek_root, curved_blend, FILTER_IGNORE, true);
 			time = p_time;
 		} else {
-			blend_input(prev, p_time, false, p_seek_root, blend, FILTER_IGNORE, true);
+			blend_input(prev, p_time, false, p_seek_root, curved_blend, FILTER_IGNORE, true);
```
Comment (TokageItLab) [inline scene/animation/animation_blend_tree.cpp:765]: Suggestion:
```cpp
	} else { // cross-fading from prev to current

		float blend = xfade_time == 0 ? 0 : (prev_xfading / xfade_time);
		if (xfade_curve.is_valid()) {
			blend = xfade_curve->interpolate(blend);
		}

		if (from_start && !p_seek && switched) { //just switched, seek to start of current

			rem = blend_input(current, 0, true, p_seek_root, 1.0 - blend, FILTER_IGNORE, true);
		} else {
			rem = blend_input(current, p_time, p_seek, p_seek_root, 1.0 - blend, FILTER_IGNORE, true);
		}

		if (p_seek) {
			blend_input(prev, p_time, true, p_seek_root, blend, FILTER_IGNORE, true);
			time = p_time;
		} else {
			blend_input(prev, p_time, false, p_seek_root, blend, FILTER_IGNORE, true);
			time += p_time;
			prev_xfading -= p_time;
			if (prev_xfading < 0) {
				set_parameter(this->prev, -1);
			}
		}
	}
```
Your interpolation calculation is wrong; you only need to overwrite blend with curve.

File: scene/animation/animation_node_state_machine.cpp [hunk 4]
```diff
@@ -420,10 +432,12 @@ double AnimationNodeStateMachinePlayback::process(AnimationNodeStateMachine *p_s
...
+		p_state_machine->blend_node(fading_from, p_state_machine->states[fading_from].node, p_time, p_seek, p_seek_root, curved_blend, AnimationNode::FILTER_IGNORE, true);
```
Comment (TokageItLab) [inline scene/animation/animation_node_state_machine.cpp:440]: Suggestion:
```cpp
	if (current_curve.is_valid()) {
		fade_blend = current_curve->interpolate(fade_blend);
	}
	float rem = p_state_machine->blend_node(current, p_state_machine->states[current].node, p_time, p_seek, p_seek_root, fade_blend, AnimationNode::FILTER_IGNORE, true);

	if (fading_from != StringName()) {
		p_state_machine->blend_node(fading_from, p_state_machine->states[fading_from].node, p_time, p_seek, p_seek_root, 1.0 - fade_blend, AnimationNode::FILTER_IGNORE, true);
	}
```
Same above.

**Review by TokageItLab — CHANGES_REQUESTED** [review]

**Comment by TokageItLab** [comment]
@jeronimo-schreyer This feature is good. Can I salvage it?
We have to do a feature freeze to release the Godot 4 beta, so implementation of the new feature is a bit rushed. If I don't hear back by 8/2, I will send a proxy PR with your name as co-authored.

**Comment by akien-mga** [comment]
Superseded by #63802.

**Pull request closed without merging**